README.md
Extract unique Python-Exceptions with their traceback from a log file.

## WARNING!

The extraction logic here is heuristic and may fail you. Don't depend on it for any life or death situations. Also, please submit feedback on how it could be improved!

## Installational

 Clone this repository and install with

    python setup.py install

This adds a utility called `exceptional` or `exceptional.exe` to your path.

## Usage

To extract the Python tracebacks from a log called _logfile.txt_, run:

    $ exceptional -f logfile.txt

If you want to exclude certain exceptions, try:

    $ exceptional -f logfile.txt -e ValueError,AttributeError

You can all pass multiple filenames:

    $ exceptional -f logfile1.txt logfile2.txt

This would exclude would exclude any ``ValueError`` or ``AttributeError`` tracebacks from the output.

The tool can also read the log file from stdout, e.g.:

    cat logfile.txt | exceptional

or

    cat logfile.txt | exceptional -e ValueError

## Attribution

This is based on [a script](https://gist.github.com/originell/1923003) by [@originell](https://github.com/originell).

HISTORY.rst
=======
History
=======

0.1.0 (2016-06-16)
------------------

* First release on PyPI.


tests/test_exceptional.py

#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import, print_function

import six
import os

from subprocess import check_output

"""
test_exception
----------------------------------

Tests for `exception` module.
"""
import pytest

from exception import exception


@pytest.fixture
def simple_traceback():
    return """Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ZeroDivisionError: integer division or modulo by zero"""


@pytest.fixture
def simple_traceback_buffer():
    return six.StringIO("""Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ZeroDivisionError: integer division or modulo by zero""")


@pytest.fixture
def mock_filename(mocker):
    mocker.patch('fileinput.filename', spec=True)
    exception.fileinput.filename.return_value = "X"


def test_trivial(mock_filename, simple_traceback):
    """Check that a simple example is fine"""
    errors = exception.extract_errors(simple_traceback.split("\n"))
    out = [error for filename, error in errors]
    assert len(out) == 1
    assert "".join(simple_traceback.split('\n')) == out[0]

    # Add some trash before and afterthe
    trace2 = "a\n{}\na\n\n\n".format(simple_traceback)

    errors = list(exception.extract_errors(trace2.split("\n")))
    out = [error for filename, error in errors]
    assert len(out) == 1
    assert "".join(simple_traceback.split('\n')) == out[0]

    assert errors[0][0] == "X"


def test_file(mock_filename, simple_traceback, simple_traceback_buffer):
    """Correctly parses an input buffer"""
    errors = exception.extract_errors(simple_traceback_buffer.readlines())
    out = [error for filename, error in errors]
    assert len(out) == 1
    assert "".join(simple_traceback) == out[0]


def test_multiple_exceptions(mock_filename, simple_traceback):
    """Extract two exceptions from a string"""
    trace1 = simple_traceback
    trace2 = simple_traceback.replace("ZeroDivisionError", "ValueError")

    traceback = "{}\n{}".format(trace1, trace2)
    errors = exception.extract_errors(traceback.split("\n"))
    out = [error for filename, error in errors]
    assert len(out) == 2
    assert "".join(trace1.split('\n')) == out[0]
    assert "".join(trace2.split('\n')) == out[1]


def test_deduplicate(mock_filename, simple_traceback):
    """Duplicate exceptions in sequence are ignored"""
    traceback = "{}\n{}".format(simple_traceback, simple_traceback)

    errors = exception.extract_errors(traceback.split("\n"))
    out = [error for filename, error in errors]
    assert len(out) == 1
    assert "".join(simple_traceback.split('\n')) == out[0]


@pytest.mark.skipif('/home/travis/' in os.environ.get('PATH', ''),
                    reason="Can't get path to work on Travis")
def test_cli():
    print(os.environ)
    root = os.path.dirname(__file__)
    result = check_output(["exception", "-f", "{}/fixtures/real_log.log".format(root)])

    true = """### {}/fixtures/real_log.log ###

Traceback (most recent call last):
  File "local/lib/python2.7/site-packages/pykafka/cluster.py", line 242, in _request_metadata
    ssl_config=self._ssl_config)
  File "local/lib/python2.7/site-packages/pykafka/broker.py", line 97, in __init__
    self.connect()
  File "local/lib/python2.7/site-packages/pykafka/broker.py", line 211, in connect
    self._connection.connect(self._socket_timeout_ms)
  File "local/lib/python2.7/site-packages/pykafka/connection.py", line 169, in connect
    (self.source_host, self.source_port)
  File "/usr/lib/python2.7/socket.py", line 571, in create_connection
    raise err
error: [Errno 111] Connection refused

----------------------------------------
""".format(root)
    assert result == true


exception/cli.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, print_function

import fileinput

from argparse import ArgumentParser

from .exception import extract_errors, print_error


def get_options():
    parser = ArgumentParser()
    parser.add_argument('-f', '--file', dest='file',
                        help='The file to extract exceptions from',
                        metavar='FILE',
                        nargs='+',
                        type=str)
    parser.add_argument('-e', '--exclude', dest='exclude_list',
                        help='Exclude certain exceptions from output.',
                        default="",
                        metavar='Exception,Exception,...')
    return parser.parse_args()


def main():
    options = get_options()
    for filename, error in extract_errors(fileinput.input(options.file),
                                          options.exclude_list):
        print("### %s ###\n" % filename)  # TODO: add a flag for this
        print_error(error)


if __name__ == "__main__":
    main()


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'tdhopper/exception'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


exceptional/exceptional.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, print_function

import fileinput

from argparse import ArgumentParser

# SETTINGS
MAX_LINE_LENGTH = 400
DELIMETER_LENGTH = 40


def extract_errors(line_generator, exclude_list=None):
    buffer_on = False
    error = ''
    last_error = ''
    for line in line_generator:
        if is_first_line(line):
            buffer_on = True
            first_line = line
        if buffer_on:
            # Truncate lines longer than 400 characters.
            if len(line) > MAX_LINE_LENGTH:
                line = line[:MAX_LINE_LENGTH]+'...\n'
            error += line

            if line and is_last_line(line, first_line):
                buffer_on = False
                if exclude_error(error, last_error, exclude_list):
                    last_error = error
                    error = ''
                    continue
                else:
                    yield fileinput.filename(), error
                    last_error = error
                    error = ''


def is_first_line(line):
    return line.startswith("Traceback (most recent call last):")


def is_last_line(line, first_line):
    return (line != first_line) and not line.startswith(" ")


def exclude_error(error, last_error, exclude_list):
    excludes = exclude_list.split(',') if exclude_list else []
    if any([excl in error for excl in excludes]):
        return True
    elif error.strip() == "":
        return True
    elif error == last_error:
        return True
    else:
        return False


def print_error(error):
    print(error)
    print(DELIMETER_LENGTH * "-")



exceptional/__main__.py
from __future__ import print_function
from argparse import ArgumentParser
import fileinput

# SETTINGS
MAX_LINE_LENGTH = 400
DELIMETER_LENGTH = 40


def get_options():
    parser = ArgumentParser()
    parser.add_argument('-f', '--file', dest='file',
                      help='The file to extract exceptions from',
                      metavar='FILE',
                      nargs='+',
                      type=str)
    parser.add_argument('-e', '--exclude', dest='exclude_list',
                      help='Exclude certain exceptions from output.',
                      default="",
                      metavar='Exception,Exception,...')
    return parser.parse_args()


def extract_errors(options):
    buffer_on = False
    error = ''
    last_error = ''
    for line in fileinput.input(options.file):
        if is_first_line(line):
            buffer_on = True
            first_line = line
        if buffer_on:
            # Truncate lines longer than 400 characters.
            if len(line) > MAX_LINE_LENGTH:
                line = line[:MAX_LINE_LENGTH]+'...\n'
            error += line
            if line and is_last_line(line, first_line):
                buffer_on = False
                if exclude_error(error, last_error, options.exclude_list):
                    last_error = error
                    error = ''
                    continue
                else:
                    yield fileinput.filename(), error
                    last_error = error
                    error = ''


def is_first_line(line):
    return line.startswith("Traceback (most recent call last):")


def is_last_line(line, first_line):
    return (line != first_line) and not line.startswith(" ")


def exclude_error(error, last_error, exclude_list):
    excludes = exclude_list.split(',') if exclude_list else []
    if any([excl in error for excl in excludes]):
        return True
    elif error.strip() == "":
        return True
    elif error == last_error:
        return True
    else:
        return False


def print_error(error):
    print(error)
    print(DELIMETER_LENGTH * "-")


def main():
    options = get_options()
    for filename, error in extract_errors(options):
        print("### %s ###\n" % filename) # TODO: add a flag for this
        print_error(error)


if __name__ == "__main__":
    main()